keps/sig-network/20200130-Uncapped-IPv6.md

```diff
@@ -0,0 +1,290 @@
+
+### Goals
+
+Replace current bitmap structures that are causing IPv6 restrictions by [roaring bitmaps](http://roaringbitmap.org/about/).
+
+### Non-Goals
+
+Replace other bitmap structures that are not causing IPv6 restrictions.
+
+Move out of the core REST or refactor the port and ip allocators.
+
+## Proposal
+
+Kubernetes has the following IPv6 restrictions:
+
+- Maximum subnet mask size in the Node IPAM Controller Range Allocator, given by the difference between the cluster-cidr-mask size and the node-cidr-mask size.
+
+https://github.com/kubernetes/kubernetes/blob/bd1195c28e22bbdcb95d718a31186581dc0a5f5f/pkg/controller/nodeipam/ipam/cidrset/cidr_set.go#L42-L51
+
+```go
+// The subnet mask size cannot be greater than 16 more than the cluster mask size
+// TODO: https://github.com/kubernetes/kubernetes/issues/44918
+// clusterSubnetMaxDiff limited to 16 due to the uncompressed bitmap
+// Due to this limitation the subnet mask for IPv6 cluster cidr needs to be >= 48
+// as default mask size for IPv6 is 64.
+clusterSubnetMaxDiff = 16
+// halfIPv6Len is the half of the IPv6 length
+halfIPv6Len = net.IPv6len / 2
+```
+
+- Maximum subnet size (/12 for IPv4 and /112 for IPv6) for the Service Cluster CIDR in the kube-apiserver:
+
+https://github.com/kubernetes/kubernetes/blob/bd1195c28e22bbdcb95d718a31186581dc0a5f5f/cmd/kube-apiserver/app/options/validation.go#L48-L54
+
+```go
+// Complete() expected to have set Primary* and Secondary*
+// primary CIDR validation
+var ones, bits = options.PrimaryServiceClusterIPRange.Mask.Size()
+if bits-ones > 20 {
+  errs = append(errs, errors.New("specified --service-cluster-ip-range is too large"))
+}
+```
+
+- Maximum number of IPs (65536 for IPv6) for the Service Cluster subnet in the service IP allocator:
+
+https://github.com/kubernetes/kubernetes/blob/5e31799701123c50025567b8534e1a62dbc0e9f6/pkg/registry/core/service/ipallocator/allocator.go#L275-L283
+
+```go
+  // For IPv6, the max size will be limited to 65536
+  // This is due to the allocator keeping track of all the
+  // allocated IP's in a bitmap. This will keep the size of
+  // the bitmap to 64k.
+  if bits == 128 && (bits-ones) >= 16 {
+    return int64(1) << uint(16)
+  } else {
+    return int64(1) << uint(bits-ones)
+  }
+```
+
+### Roaring bitmaps
+
+Roaring bitmaps are compressed bitmaps which tend to outperform conventional compressed bitmaps such as WAH, EWAH or Concise. In some instances, they can be hundreds of times faster and they often offer significantly better compression.
+
+There is a [roaring bitmap go version](https://github.com/RoaringBitmap/roaring) that will be imported into the Kubernetes project with its dependencies.
+
+- github.com/RoaringBitmap/roaring
+- github.com/glycerine/go-unsnap-stream
+- github.com/glycerine/goconvey
+- github.com/golang/snappy
+- github.com/mschoch/smat
+- github.com/philhofer/fwd
+- github.com/tinylib/msgp
+- github.com/willf/bitset
```

Comment:
is this the full dependency set (including all transitives)?

Reply:
I took it from here https://github.com/kubernetes/kubernetes/pull/86620/files